Add Twirp Design Principles to the Contributing Guidelines and archive protocol v6 (#237)
twitchtv/twirp: CONTRIBUTING.md

# Contributing

Thanks for helping make Twirp better! This is great!

## Twirp Principles

Your contribution will go more smoothly if it aligns well with the project priorities. Twirp has been developed by twitch.tv and it has been in production for years. As many internal systems, teams and processes depend on Twirp, keeping backwards compatibility is taken very seriously.

Design principles:

 * Simplicity. Keep serialization and routing simple and intuitive.
 * Small api and interface. Reduce future support burden.
 * Prevent user errors. Avoid surprising behavior.
 * Pragmatism over being bleeding edge.
 * Hooks are for observability instead of control flow, so it is easier to debug Twirp systems.
 * No flags/options for code generation, so generated code is predictable across versions and platforms.
 * As few dependencies as possible, so it is easier to integrate and upgrade.
 * Prefer generated code over shared libraries between services and clients, so it is easier to implement changes without breaking compatibility.

Contributions that are welcome:

 * Security updates.
 * Performance improvements.

>  * Supporting new versions of Golang and Protobug.

Reviewer: ```suggestion
 * Supporting new versions of Golang and Protobuf.
```

 * Documentation.
 * Making Twirp easier to integrate with other useful tools.

> In the other hand, contributions that contradict the following priorities will be more difficult:

Reviewer: ```suggestion
On the other hand, contributions that contradict the following priorities will face more resistance:
```

 * Complex and involved features should be avoided. Twirp was designed to be easy to use and easy to maintain. See the streaming proposal as an example of a complex feature that was not able to make it into a major release: https://github.com/twitchtv/twirp/issues/3.
 * Features that could be implemented in a separate library should go in a separate library. Twirp allows easy integrtion with 3rd party code.

## Report an Issue

If you have run into a bug or want to discuss a new feature, please [file an issue](https://github.com/twitchtv/twirp/issues). If you'd rather not publicly discuss the issue, please email security@twitch.tv.

## Contributing Code with Pull Requests

Twirp uses github pull requests. Fork, hack away at your changes and submit. Most pull requests will go through a few iterations before they get merged. Different contributors will sometimes have different opinions, and often patches will need to be revised before they can get merged.

### Requirements

 * Twirp officially supports the last 3 releases of Go.

>  * The Python implementation uses Python 2.7.

Reviewer: Python has stopped official support for `2.7`, maybe remove the requirement?

https://github.com/python/devguide/pull/344

Author: That is right. However this should be discussed in a different PR. The Python implementation has a few things that should be discussed all together. This PR adds some edits to the Contributing Guidelines, but with the focus on adding the Twirp Principles.

Author: We edited this point with: "As such, it is not suitable for production use in its current form.". Thanks for the reference.
